src/tokens/reader_stream.rs

```diff
@@ -0,0 +1,560 @@
+    scalar_content_method!(timestamp_token, Timestamp, reader, reader.read_timestamp()?);
+    scalar_content_method!(string_token, String, reader, reader.read_string()?);
+    scalar_content_method!(symbol_token, Symbol, reader, reader.read_symbol()?);
+    scalar_content_method!(blob_token, Blob, reader, reader.read_blob()?.into());
+    scalar_content_method!(clob_token, Clob, reader, reader.read_clob()?.into());
+
+    #[inline]
+    fn invalidate_token(&mut self) {
+        // only invalidate if we have something deferred
+        if self.pending_token.is_deferred() {
+            let underlying = {
+                let mut original = self.reader_cell.borrow_mut();
+                original.invalidate()
+            };
+            self.reader_cell = Rc::new(RefCell::new(underlying));
+        }
+    }
+
+    #[inline]
+    fn next(&mut self) -> IonResult<StreamItem> {
+        let mut underlying = self.reader_cell.borrow_mut();
+        let reader = underlying.as_mut_ref();
+        self.position = ReaderPosition::Nothing;
+        let item = reader.next()?;
+        self.position = match item {
+            StreamItem::Value(ion_type) => match ion_type {
+                IonType::List => ReaderPosition::Container(ContainerType::List),
+                IonType::SExp => ReaderPosition::Container(ContainerType::SExp),
+                IonType::Struct => ReaderPosition::Container(ContainerType::Struct),
+                _ => ReaderPosition::NonContainer,
+            },
+            StreamItem::Null(_) => ReaderPosition::NonContainer,
+            StreamItem::Nothing => ReaderPosition::Nothing,
+        };
+        Ok(item)
+    }
+
+    #[inline]
+    fn step_in(&mut self) -> IonResult<()> {
+        let mut underlying = self.reader_cell.borrow_mut();
+        let reader = underlying.as_mut_ref();
+        self.position = ReaderPosition::Nothing;
+        reader.step_in()
+    }
+
+    #[inline]
+    fn step_out(&mut self) -> IonResult<()> {
+        let mut underlying = self.reader_cell.borrow_mut();
+        let reader = underlying.as_mut_ref();
+        self.position = ReaderPosition::Nothing;
+        reader.step_out()
+    }
+
+    #[inline]
+    fn is_null(&self) -> bool {
+        let underlying = self.reader_cell.borrow();
+        let reader = underlying.as_ref();
+        reader.is_null()
+    }
+
+    #[inline]
+    fn ion_type(&self) -> Option<IonType> {
+        let underlying = self.reader_cell.borrow();
+        let reader = underlying.as_ref();
+        reader.ion_type()
+    }
+
+    #[inline]
+    fn parent_type(&self) -> Option<IonType> {
+        let underlying = self.reader_cell.borrow();
+        let reader = underlying.as_ref();
+        reader.parent_type()
+    }
+}
+
+impl<'a, R> TokenStream<'a> for ReaderTokenStream<'a, R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol> + 'a,
+{
+    fn next_token(&mut self, instruction: Instruction) -> IonResult<Token<'a>> {
```

Comment:
🗺️ Originally this was tied to the borrow of `&mut self`.  The problem with this is that you _must_ materialize the token if you want to _implement_ `IonReader` over `IonStream` (see #542).  This is a problematic design since having an `IonReader` that has to materialize the token on every call to `next` is not efficient.